src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs

```diff
@@ -363,122 +384,174 @@ private TimeSpan RunTestsInternal()
+                    // PlatformServices.Default.Runtime.RuntimeVersion);
+                    var runtimeVersion = " ";
+                    this.TestRunEventsHandler?.HandleLogMessage(
+                        TestMessageLevel.Warning,
+                        string.Format(
+                            CultureInfo.CurrentUICulture,
+                            CrossPlatEngineResources.NoMatchingExecutor,
+                            executorUriExtensionTuple.Item1.AbsoluteUri,
+                            runtimeVersion));
 
-                        // If test run cancellation is requested, skip the next executor
-                        if (this.isCancellationRequested)
-                        {
-                            break;
-                        }
+                    continue;
+                }
 
-                        var timeStartNow = DateTime.UtcNow;
+                // Cache the executor.
+                executorCache.Add(executorUriExtensionTuple.Item1.AbsoluteUri, executor);
 
-                        var currentTotalTests = this.testRunCache.TotalExecutedTests;
-                        this.testPlatformEventSource.AdapterExecutionStart(executorUriExtensionTuple.Item1.AbsoluteUri);
+                // Check if we actually have to attach to the default test host.
+                if (!this.runContext.IsBeingDebugged || attachedToTestHost)
+                {
+                    // We already know we should attach to the default test host, simply continue.
+                    continue;
+                }
 
-                        // Run the tests.
-                        if (this.NotRequiredSTAThread() || !this.TryToRunInSTAThread(() => this.InvokeExecutor(executor, executorUriExtensionTuple, this.runContext, this.frameworkHandle), true))
-                        {
-                            this.InvokeExecutor(executor, executorUriExtensionTuple, this.runContext, this.frameworkHandle);
-                        }
+                // If there's at least one adapter in the filtered adapters list that doesn't
+                // implement the new test executor interface, we should attach to the default test
+                // host by default.
+                // Same goes if all adapters implement the new test executor interface but at
+                // least one of them needs the test platform to attach to the default test host.
+                if (!(executor.Value is ITestExecutor2)
+                    || this.ShouldAttachDebuggerToTestHost(executor, executorUriExtensionTuple, this.runContext))
+                {
+                    EqtTrace.Verbose("Attaching to default test host.");
 
-                        this.testPlatformEventSource.AdapterExecutionStop(this.testRunCache.TotalExecutedTests - currentTotalTests);
+                    attachedToTestHost = true;
+                    var pid = Process.GetCurrentProcess().Id;
+                    if (!this.frameworkHandle.AttachDebuggerToProcess(pid))
```

Comment:
>AttachDebuggerToProcess [](start = 46, length = 23)

Do we already have telemetry that logs the protocol versions across vstest.console, testhost and in this case how many adapters implement ITestExecutor2? It would be useful for us to understand usage stats.

Reply:
I'll open another follow-up task to add telemetry.